Allow required options to be satisfied by env var

gwatts opened the pull request:
If an option is required, but omitted from the argument list,
allow it to fallback to a configured environment variable instead.

Continue to disallow falling back to a default value - ie. the end user
must either supply the option on the command line, or set the
environment variable.


--

This is a rebase of #34 and adds a unit test... probably a better or more comprehensive way to write that test; let me know if so.

jawher commented:
Thanks @gwatts for having taken the time to rework your initial contribution after all the breaking changes I made.

I'm in Brussels (for the FOSDEM) this weekend but I'll be reviewing your PR first thing after coming back.

gwatts commented:
A slight fly in the ointment is that errors from custom types are not reported if they're set by environment variable.. Consider:

```golang
package main

import (
        "fmt"
        "os"
        "strings"

        "github.com/jawher/mow.cli"
)

type OneOfString struct {
        Choices []string
        Value   string
}

func (so *OneOfString) Set(v string) error {
        for _, c := range so.Choices {
                if c == v {
                        so.Value = v
                        return nil
                }
        }
        return fmt.Errorf("Invalid selection; must be one of %s", strings.Join(so.Choices, ","))
}

func (so *OneOfString) String() string {
        return so.Value
}

func NewOneOfString(s ...string) *OneOfString {
        return &OneOfString{Choices: s}
}

func main() {
        app := cli.App("test", "test")

        val := NewOneOfString("foo", "bar")

        app.Spec = "--oneof"
        app.Var(cli.VarOpt{
                Name:   "oneof",
                Desc:   `Must be either "foo" or "bar"`,
                EnvVar: "ONEOF",
                Value:  val,
        })

        app.Action = func() {
                fmt.Println("Got value: ", val.Value)
        }

        app.Run(os.Args)
}
```

```
$ go run ./env.go --oneof=badvalue
Error: Invalid selection; must be one of foo,bar

Usage: test --oneof

test

Options:
  --oneof=     Must be either "foo" or "bar" ($ONEOF)
exit status 2
```

which correctly outputs the custom types `Error: Invalid selection` error response

vs.

```
$ ONEOF=badvalue go run ./env.go 
Error: incorrect usage

Usage: test --oneof

test

Options:
  --oneof=     Must be either "foo" or "bar" ($ONEOF)
exit status 2
```

which just shows incorrect usage with no further help.

and of course (for this PR), supplying a good value by environment variable does the right thing:

```
$ ONEOF=foo go run ./env.go 
Got value:  foo
```

If there's a good way to address that with this PR then i can take a look, else I can open another issue.

jawher commented:
@gwatts I don't think this is an issue, or at least it's a tiny one.

This doesn't change the old behaviour, i.e. silently ignoring invalid values in the env vars.

jawher commented:
@gwatts So, I reviewed this PR and I have one bit of feedback: while I like the coverage of your test, I think it's a bit too complex, and I had/have a hard time reading it and trying to understand what's going on.

Could you consider simplifying it a bit:

* Splitting the OK/KO cases in two separate tables, e.g. ` okCases`  and `badCases`  ? This will get rid of a bool values (in the cases) and an if/else branch.
* What is the purpose of the string arg `ARG` ?

gwatts commented:
@jawher thanks for the review; i've split the test into two, which simplifies thing a bit and removed the `ARG` string argument, which i think was a remnant of a previous test

jawher commented:
@gwatts Thank you for your contribution !

gwatts commented:
np - thanks for a great CLI library! 👍
